File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/Drivetrain.java
package org.firstinspires.ftc.teamcode.hardware;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.Gamepad;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.teamcode.Subsystem;
import org.jetbrains.annotations.Nullable;

public class Drivetrain implements Subsystem {
    private static final double SCALE_FACTOR = -0.8;
    private static final int[][] MULTIPLIERS = {
            {+1, +1},
            {+1, -1},
            {-1, +1},
            {-1, -1}
    }; // z doesn't need a multiplier since everything is +1
    private final DcMotor[] all;
    private final Gamepad gamepad;

    public Drivetrain(HardwareMap hardwareMap, @Nullable Gamepad gamepad) {
        this.gamepad = gamepad;
        all = new DcMotor[] {
                hardwareMap.get(DcMotor.class, "frontLeft"),
                hardwareMap.get(DcMotor.class, "frontRight"),
                hardwareMap.get(DcMotor.class, "backLeft"),
                hardwareMap.get(DcMotor.class, "backRight")
        };
    }

    @Override
    public void loop() {
        double x = -gamepad.left_stick_x;
        double y = gamepad.left_stick_y;
        double z = -gamepad.right_stick_x;

        double total = Math.abs(x) + Math.abs(y) + Math.abs(z);

        if (total == 0) { // prevent division by 0
            for (DcMotor motor : all) motor.setPower(0);
            return;
        }

        // Adjust input to never exceed 1
        double factor = Math.max(Math.hypot(x, y), Math.abs(z)) * SCALE_FACTOR / total;
        for (int i = 0; i < 4; i++) all[i].setPower(
                ((MULTIPLIERS[i][0] * x) + (MULTIPLIERS[i][1] * y) + z) * factor
        );
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/Drivetrain.kt
package org.firstinspires.ftc.teamcode.hardware

import com.qualcomm.robotcore.eventloop.opmode.OpMode
import org.firstinspires.ftc.teamcode.Subsystem
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.Gamepad
import com.qualcomm.robotcore.util.ElapsedTime
import org.firstinspires.ftc.teamcode.recording.InputSequence
import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.pow

private typealias Scaling = (Float, Float, Float) -> Double

class Drivetrain(private val opMode: OpMode, private val gamepad: Gamepad = opMode.gamepad1) : Subsystem {
    private val multiplierMap: Map<DcMotor, MotorMultipliers>
//    private val ordometry: Array<DcMotor>
    private val sequence: InputSequence = InputSequence()
    private val timer = ElapsedTime()

    init {
        val hardwareMap = opMode.hardwareMap
        multiplierMap = hashMapOf(
            hardwareMap.dcMotor["frontLeft"] to MotorMultipliers(+1, +1),
            hardwareMap.dcMotor["frontRight"] to MotorMultipliers(+1, -1),
            hardwareMap.dcMotor["backLeft"] to MotorMultipliers(-1, +1),
            hardwareMap.dcMotor["backRight"] to MotorMultipliers(-1, -1)
        ) // z doesn't need a multiplier since everything is +1

//        ordometry = arrayOf(
//            hardwareMap.dcMotor["encoderX"],
//            hardwareMap.dcMotor["encoderY"],
//        )

        multiplierMap.keys.forEach {
            it.mode = DcMotor.RunMode.RUN_USING_ENCODER
            it.direction = DcMotorSimple.Direction.FORWARD
        }
    }
    // Map()


    @Suppress("NAME_SHADOWING")
    fun move(x: Float = 0f, y: Float = 0f, z: Float = 0f,
             scaleFunction: Scaling = TELEOP_SCALING) {
        val total = abs(x) + abs(y) + abs(z)

        if (total == 0f) return stop() // prevent division by 0

        setMotorPowers { _, mults -> (mults.x * x + mults.y * y + z) * scaleFunction(x, y, z) * SCALE_FACTOR / total }
    }

    private fun stop() = multiplierMap.keys.forEach { it.power = 0.0 }

    override fun loop() {
        if (gamepad.dpad_right) return setMotorPowers { _, mults -> 0.5 * mults.y * SCALE_FACTOR }
        if (gamepad.dpad_left) return setMotorPowers { _, mults -> -0.5 * mults.y * SCALE_FACTOR }

        move(
            x = gamepad.left_stick_x,
            y = -gamepad.left_stick_y, // for some atrocious reason, down is positive
            z = gamepad.right_stick_x
        )

//        opMode.telemetry.addData("encoderX", odometry[0].currentPosition)
//        opMode.telemetry.addData("encoderY", odometry[1].currentPosition)
    }

    data class MotorMultipliers(val x: Int, val y: Int)

    private inline fun setMotorPowers(powerGetter: (DcMotor, MotorMultipliers) -> Double) {
        multiplierMap.forEach { (motor, mults) ->
            motor.power = powerGetter(motor, mults)
        }
    }

    fun setPower(x: Double, y: Double, z: Double, scaling: Scaling = TELEOP_SCALING) {
        move(x.toFloat(), y.toFloat(), z.toFloat(), scaling)
    }

    companion object {
        const val SCALE_FACTOR = 1.0
        val TELEOP_SCALING: Scaling = { x, y, z -> max(0.2 , max(hypot(x, y), abs(z)).pow(2f).toDouble()) }
        val LINEAR_SCALING: Scaling = { x, y, z -> max(hypot(x, y), abs(z)).toDouble() }
    }
}
